src/components/Button/Button.tsx

```diff
@@ -124,6 +126,7 @@ export function Button({
 }: ButtonProps) {
   const {unstableGlobalTheming = false} = useFeatures();
   const hasGivenDeprecationWarning = useRef(false);
+  const id = useUniqueId('Button', idProp);
```

Comment:
Having to add an ID to every button feels rather heavy handed, I wonder if there is a way we can avoid this (or if not, can we make the id needed by the context be distinct from the id attribute that gets added to the DOM).

Perhaps instead of having the Buttons be responsible for adding themselves to the context, instead ButtonGroup could trawl through its children to collect the Button elements and adjust their props?

Reply:
What do you think of just using the [Id internally](https://github.com/Shopify/polaris-react/pull/2441/files#diff-80ce11d837170156710fe9e24d078a78R129)? 

Having ButtonGroup look at its children and add props I think could be done by recursively looping through the children and adding a prop to the Button by using cloneElement. Is that what you have in mind? Doing that would limit our implementation to Buttons. Having Buttons register themselves means this can be added to any component using the registration hook. For instance, the first button in BulkAction isn't a Button per se, but with this approach, we can still make use to the group.

Reply:
I'd be on board with using the id internally, I think that pollutes the DOM a little less.

As an aside how many renders does this approach trigger? Do we pay a price for rendering the button group, then when rendering the button updating the context (does that cause another render)

> Having ButtonGroup look at its children and add props I think could be done by recursively looping through the children and adding a prop to the Button by using cloneElement. Is that what you have in mind? Doing that would limit our implementation to Buttons. Having Buttons register themselves means this can be added to any component using the registration hook.

Something like that yeah, though I'm not sure about its performance characteristics so I don't know if it's a good idea or not. We're changing the registration point from within the child element (e.g. Button) to the container (i.e. ButtonGroup) - we can make that say "look for Buttons and Popovers". We wouldn't be able to account for any component (e.g. something that exists in web) but being able to open that up to ANYTHING can be a child of a ButtonGroup feels a bit too open-ended to me at the moment (though I've not thought about this in depth)

Reply:
> As an aside how many renders does this approach trigger? Do we pay a price for rendering the button group, then when rendering the button updating the context (does that cause another render)

That is a bit of a concern that it renders every time a button renders but the render is negligible I would say because buttons are small and there won't be that many in a group.